proposals/XXXX-efficient-token-program.md

```diff
@@ -0,0 +1,78 @@
+---
+simd: 'XXXX'
+title: Efficient Token program
+authors:
+  - (fill in with names of authors)
+category: Standard
+type: Core
+status: Draft
+created: 2025-01-26
+feature: (fill in with feature tracking issues once accepted)
+supersedes: (optional - fill this in if the SIMD supersedes a previous SIMD)
+superseded-by: (optional - fill this in if the SIMD is superseded by a subsequent
+ SIMD)
+extends: (optional - fill this in if the SIMD extends the design of a previous
+ SIMD)
+---
+
+## Summary
+
+Replace the current version of SPL Token (`TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA`) program by a CU-optimized one (`p-token`).
+
+## Motivation
+
+About `~10%` of block compute units is used by the Token program instructions. Decreasing the CUs used by Token program instruction creates block space for other transactions to be executed, i.e., less CUs consumed by Token program, more CUs for other transactions.
+
+As an example, if we reduce the CUs consumed by Token instructions to `1/20th` of their current value, `10%` of block CUs utilized becomes `0.5%`, resulting in a `9.5%` block space gain.
+
+Additionally, there are benefits to downstream programs:
+* Better composability since using the Token program instructions require less CUs.
+* Cheaper (in terms of CU) cross-program invocations.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
```

Comment:
Under the detailed design, we need to spell out the work to be done by validator implementations.

Specifically, when the accompanying feature is enabled, the runtime must replace the contents of the account at `Tokenkeg...` with the contents of `$OTHER_ACCOUNT_ADDRESS`. 

But since we can't use BPF Loader v2 anymore, we'll need to swap `Tokenkeg...` with a program using the upgradeable loader. This means that we'll also need to replace the contents of `$TOKENKEG_PROGRAM_DATA_ADDRESS` with the contents of `$OTHER_ACCOUNT_PROGRAM_DATA_ADDRESS`, where the program data address is defined as `PDA([program_id], upgradeable_loader_id)`.

Code reference at https://github.com/anza-xyz/agave/blob/74b92ed36a70026629a0b303c2eeeff1e8e49f86/sdk/loader-v3-interface/src/lib.rs#L26

Reply:
The current `Tokenkeg...` does not have a `$TOKENKEG_PROGRAM_DATA_ADDRESS`, right? The program data is in the actual `Tokenkeg...` account since it is a BPF Loader v2 program account. 

If that is the case, then we can leave the `$OTHER_ACCOUNT_PROGRAM_DATA_ADDRESS` as it is, and replace `Tokenkeg...` contents with `$OTHER_ACCOUNT_ADDRESS` – the new one will already be referencing the `$OTHER_ACCOUNT_PROGRAM_DATA_ADDRESS`.

Reply:
Oooh, I see, `$OTHER_ACCOUNT_PROGRAM_DATA_ADDRESS` won't have the correct derivation.